packages/@eventual/aws-cdk/src/socket-service.ts

```diff
@@ -0,0 +1,123 @@
+import { ServiceLocal } from "./service.js";
+import { ServiceEntityProps } from "./utils.js";
+import { DeepCompositePrincipal } from "./deep-composite-principal.js";
+
+export type ApiOverrides = Omit<SpecHttpApiProps, "apiDefinition">;
+
+export type Sockets<Service> = ServiceEntityProps<Service, "Socket", Socket>;
+
+export type SocketOverrides<Service> = Partial<
+  ServiceEntityProps<Service, "Socket", SocketHandlerProps>
+>;
+
+export interface SocketsProps<Service = any>
+  extends WorkerServiceConstructProps {
+  local: ServiceLocal | undefined;
+  openApi: {
+    info: openapi.InfoObject;
+  };
+  overrides?: SocketOverrides<Service>;
+}
+
+/**
+ * Properties that can be overridden for an individual API handler Function.
+ */
+export type SocketHandlerProps = Partial<
+  Omit<FunctionProps, "code" | "runtime" | "functionName">
+>;
+
+export class SocketService<Service = any> {
+  /**
+   * API Gateway for providing service api
+   */
+  public readonly sockets: Sockets<Service>;
+
+  constructor(props: SocketsProps<Service>) {
+    const socketsScope = new Construct(props.serviceScope, "Sockets");
+
+    this.sockets = Object.fromEntries(
+      Object.entries(props.build.sockets).map(([name, socket]) => [
+        name,
+        new Socket(socketsScope, {
+          serviceProps: props,
+          socketService: this,
+          socket,
+          local: props.local,
+        }),
+      ])
+    ) as Sockets<Service>;
+  }
+}
+
+interface SocketProps {
+  serviceProps: SocketsProps<any>;
+  socketService: SocketService<any>;
+  socket: SocketFunction;
+  local: ServiceLocal | undefined;
+}
+
+class Socket extends Construct implements EventualResource {
+  public grantPrincipal: IPrincipal;
+  public gateway: IWebSocketApi;
+  public handler: Function;
+
+  constructor(scope: Construct, props: SocketProps) {
+    const socketName = props.socket.spec.name;
+
+    super(scope, socketName);
+
+    this.handler = new ServiceFunction(this, "DefaultHandler", {
+      build: props.serviceProps.build,
+      bundledFunction: props.socket,
+      functionNameSuffix: `socket-${socketName}-default`,
+      serviceName: props.serviceProps.serviceName,
+      defaults: {
+        timeout: Duration.minutes(1),
+        environment: {
+          [ENV_NAMES.SOCKET_NAME]: socketName,
+          ...props.serviceProps.environment,
+        },
+      },
+      runtimeProps: props.socket.spec.options,
+      overrides: props.serviceProps.overrides?.[socketName],
+    });
+
+    const integration = new WebSocketLambdaIntegration("default", this.handler);
+
+    this.gateway = new WebSocketApi(this, "Gateway", {
+      apiName: socketServiceSocketName(
+        props.serviceProps.serviceName,
+        socketName
+      ),
+      defaultRouteOptions: {
+        integration,
+      },
+      connectRouteOptions: {
+        integration,
+      },
+      disconnectRouteOptions: {
+        integration,
+      },
+    });
```

Comment:
There should only ever be one of these per Service.

Is this possible?

```ts
socket("foo")// ws://../foo
socket("bar") // ws://../bar
```

We don't want too many endpoints to know about.

Reply:
Not the way that websockets work, they have a single url, with a single $connect and $disconnect then route with the contents of the payload (routeKey/action), falling back to `$default`.

Conversation from Discord for context

Sam Sussman — Today at 11:14 PM
alright, socket/multiple sockets
was thinking about the routing issue
two options I can think of...

1. one web socket API + provided client - can route to the $connect handler for each based on a query string, can route to $default handler or a specific routeKey handler (later) using two json values in a managed envelope, $disconnect doesn't support routing, there can only be one... a provided client will be required and all data must be a valid json string to contain the envelope with routing keys
1. many web socket APIs + optional client - each API will have it's own $connect,  $disconnect, and $default handlers with optional handlers for route keys (later). Each API will also need to have it's own websocket url that we can expose via the cli. There will be no required envelope, except maybe to compute the optional route keys.

I am currently implementing 2 and a user could implement 1 if they were motivated to have a single url.

sam — Today at 11:17 PM
Sounds good. Simplest solution is best just to unblock yjs